### PR TITLE
Support unsigned integers in  `unwrap_cast_in_comparison` Optimizer rule

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/4148~

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3702

# Rationale for this change

This will make predicate evaluation on unsigned columns faster in some cases

# What changes are included in this PR?
1. Add support for unsigned columns
2. Add tests (using the harness I wrote in https://github.com/apache/arrow-datafusion/pull/4147)

# Are these changes tested?

Yes
# Are there any user-facing changes?

Predicates on unsigned columns should be faster

### CODE REVIEW

File: datafusion/core/tests/sql/joins.rs [hunk 1]
```diff
@@ -1428,9 +1428,9 @@ async fn reduce_left_join_1() -> Result<()> {
         "Explain [plan_type:Utf8, plan:Utf8]",
         "  Projection: t1.t1_id, t1.t1_name, t1.t1_int, t2.t2_id, t2.t2_name, t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
         "    Inner Join: t1.t1_id = t2.t2_id [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
-        "      Filter: CAST(t1.t1_id AS Int64) < Int64(100) [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
```
Comment (alamb) [inline datafusion/core/tests/sql/joins.rs:1431]: 🎉  the casts have been removed!

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 2]
```diff
@@ -292,6 +296,25 @@ fn is_support_data_type(data_type: &DataType) -> bool {
...
+    matches!(
+        dt,
+        DataType::UInt8 | DataType::UInt16 | DataType::UInt32 | DataType::UInt64
+    )
+}
+
+/// Until https://github.com/apache/arrow-rs/issues/1043 is done
+/// (support for unsigned <--> decimal casts) we also don't do that
+/// kind of cast in this optimizer
+fn is_unsupported_cast(dt1: &DataType, dt2: &DataType) -> bool {
```
Comment (alamb) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:313]: I found this with the test (which failed when it tried to invoke the arrow cast kernels for decimal <--> unsigned)

**Comment by alamb** [comment]
Here is the next PR for unwrapping casts for your review @liukun4515

File: datafusion/core/tests/sql/joins.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sql/joins.rs:1431]: This was the original regression noted in https://github.com/apache/arrow-datafusion/pull/3699#discussion_r986810683 that lead to https://github.com/apache/arrow-datafusion/issues/3702

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:313]: Casting for `unsigned value <-> decimal ` is not supported now.

Comment (liukun4515) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:313]: I will try to implement them recently.

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 3]
```diff
@@ -639,12 +693,12 @@ mod tests {
         // but the type of c6 is uint32
         // the rewriter will not throw error and just return the original expr
         let schema = expr_test_schema();
-        let expr_input = cast(col("c6"), DataType::Int64).eq(lit(0i64));
+        let expr_input = cast(col("c6"), DataType::Float64).eq(lit(0f64));
```
Comment (liukun4515) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:696]: the comment of `"c6 > 0" will be cast to `cast(c6 as int64) > 0` should be change to `"c6 > 0" will be cast to `cast(c6 as float64) > 0`

**Review by liukun4515 — APPROVED** [review]
LGTM @alamb  Thanks for your test
There is a bit comments.

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:313]: https://github.com/apache/arrow-rs/pull/3106

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 3 above)
Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:696]: Fixed in c6946f065

**Pull request merged**